jax/_src/nn/functions.py

```diff
@@ -765,3 +768,204 @@ def hard_silu(x: ArrayLike) -> Array:
+    _pv_dot_general: Callable[..., ArrayLike] = lax.dot_general) -> Array:
+  r"""Scaled dot product attention function.
+
+  Computes the attention function on Query, Key, and Value tensors:
+
+  .. math ::
+    \mathrm{Attention}(Q, K, V)=\mathrm{softmax}(\frac{QK^T}{\sqrt{d_k}}V)
+
+  If we define :code:`logits` as the output of :math:`QK^T` and the
+  :code:`probs` as the output of :math:`softmax`. This function supports
+  optional or custom (1) scale function over :code:`logits`; (2) attention mask
+  over scaled :code:`logits`; (3) softmax over masked :code:`logits`; (4)
+  dropout over :code:`probs`.
+
+  Throughout this function, we utilize the following uppercase letters to
+  represent the shape of array:
+
+    B = batch size
+    S = length of the key/value (source)
+    T = length of the query (target)
+    N = number of attention heads
+    H = dimensions of each attention head
+
+  Args:
+    query: query array; shape :code:`(BTNH)`
+    key: key array; shape :code:`(BSNH)`
+    value: value array; shape :code:`(BSNH)`
+    bias: optional, bias array to be added to logits; shape broadcastable to
+          :code:`(BNTS)`.
+    mask: optional, mask array used to filter out logits. We use additive
+          masking: 0 represents true and values below a large negative number (
+          e.g. :code:`-0.5 * jnp.finfo(dtype).max`); shape broadcastable to
+          :code:`(BNTS)`.
+    scale: scale for the logits. If None, the scale will be set to 1 divided by
+           the square root of query size.
+    logits_cap: caps the logits if it is > 0.0. Defaults to None, meaning no
+                capping. Otherwise, it applies
+                :code:`cap * jnp.tanh(logits / cap)`.
+    mask_fn: custom masking function takes in :code:`logits` and :code:`mask`
+             and outputs masked :code:`logits`.
+    softmax_fn: custom softmax function takes in :code:`logits` and outputs
+                :code:`probs`.
+    dropout_fn: custom dropout function takes in :code:`probs` outputs the new
+                `probs`.
+    _qk_dot_general: custom dot function for qk dot (mainly for quantization).
+    _pv_dot_general: custom dot function for pv dot (mainly for quantization).
+
+  Returns:
+    An array with the same shape of :code:`query`. If flash attention is not
+    used, the `probs` array will also be returned.
+  """
+  def _assert_has_shape(t: ArrayLike, shape: Sequence[int]) -> None:
+    assert t.ndim == len(shape)
+    value_str1 = f't.shape={t.shape}'
+    value_str2 = f'shape={shape}'
+    for i in range(t.ndim):
+      if shape[i] != -1:
+        assert (
+            t.shape[i] == shape[i]
+        ), f'Shapes are not equal {value_str1} vs {value_str2}'
+    
+  B, S, N, H = key.shape
+  _assert_has_shape(value, [B, S, N, H])
+  _assert_has_shape(query, [B, -1, N, H])
+  T = query.shape[1]
+  scale_val = (1.0 / np.sqrt(H) if scale is None else scale)
+
+  if not (mask is None or mask_fn is None):
```

Comment:
This further shows the limitations of the causal API: this excludes the user using fast causal attention + having a custom array mask (eg to mask per-example padding).

Reply:
For this padding + causal, we would like users to use our `MaskType.PADDING_CAUSAL` mode, which needs two additional arguments of [there](https://github.com/google/jax/blob/main/jax/_src/cudnn/fused_attention_stablehlo.py#L932-L933).

So, I am imagining like the CausalMask, we also provide `PaddingCausalMask` (not in this PR, but a followup PR probabaly)
```python
class PaddingCausalMask:
  q_seqlen: Array = None
  kv_seqlen: Array = None
```
Then, we can extract these arrays and pass them to the cudnn_sdpa().

In other words, to use cudnn sdpa, we don't recommend the mixed static mask + runtime mask, although we could do that. I think I added the quoted `mask is None or mask_fn is None` from pytorch impl to be more conservative. But if you prefer supporting such mixed usage. I can make a change.

Reply:
I think query and key sequence lengths should be arguments to the main function. Again, padding here is an implementation detail. I assume cuDNN can get speedups knowing the sequence lengths by avoiding doing unnecessary computation?

Reply:
> I think query and key sequence lengths should be arguments to the main function
With this and the is_causal argument, you can call the appropriate cuDNN padding APIs.

Reply:
Padding will be added in the next PRs.